src/fvector/fvector-perturb.cc

```diff
@@ -0,0 +1,148 @@
+  std::vector<BaseFloat> volume_factors;
+  for (MatrixIndexT i = 0; i < chunk->NumRows(); ++i) {
+    BaseFloat factor = static_cast<BaseFloat>(
+        RandInt((int)((1-opts_.max_volume_variance)*100),
+                (int)((1+opts_.max_volume_variance)*100)) / 100.0);
+    volume_factors.push_back(factor);
+  }
+  //2. scale each line respectively.
+  for (MatrixIndexT i = 0; i < chunk->NumRows(); ++i) {
+    chunk->Row(i).Scale(volume_factors[i]);
+  }
+}
+
+// we stretch the signal from the beginning to end.
+// y(t) = x(s*t) for t = 0,...,n. If s>0, the output will be shorter than
+// input. It represents speeding up. Vice versa.
+// Use ArbitraryResample deal with each line.
+//
+// In ArbitraryResample, according to num_zeros and filter_cutoff, it generates
+// the "filter_with". And then each output_sample(t) corresponds to few input_samples
+// from (t-filter_with) to (t+filter_with), which is stored in "first_index_".
+// And "weights_" will be adjust by a Hanning window in function FilterFunc.
+// In brief, you can think each output sample is the weighted sum of few input_samples.
+void FvectorPerturb::SpeedPerturbation(VectorBase<BaseFloat>& input_vector,
+                                       BaseFloat samp_freq,
+                                       BaseFloat speed_factor,
+                                       VectorBase<BaseFloat>* output_vector) {
+  if (speed_factor == 1.0) {
+    output_vector->CopyFromVec(input_vector);
+  } else {
+    Vector<BaseFloat> in_vec(input_vector),
+                      out_vec(output_vector->Dim());
+    int32 input_dim = in_vec.Dim(),
+          output_dim = out_vec.Dim();
+    Vector<BaseFloat> samp_points_secs(output_dim);
+    int32 num_zeros = 4; // Number of zeros of the sinc function that the window extends out to.
+    // lowpass frequency that's lower than 95% of the Nyquist.
+    BaseFloat filter_cutoff_hz = samp_freq * 0.475; 
+    for (int32 i = 0; i < output_dim; i++) {
+      samp_points_secs(i) = static_cast<BaseFloat>(speed_factor * i / samp_freq);
+    }
+    ArbitraryResample time_resample(input_dim, samp_freq,
+                                    filter_cutoff_hz, 
+                                    samp_points_secs,
+                                    num_zeros);
+    time_resample.Resample(in_vec, &out_vec);
+    output_vector->CopyFromVec(out_vec);
+  }
+}
+
+void FvectorPerturb::TimeShift(VectorBase<BaseFloat>& input_vector,
+                               VectorBase<BaseFloat>* output_vector) {
+  //1. generate start point randomly whose range is
+  // [0, row.NumCols()- expected_chunk_length * sample_frequency)
+  int32 start_point = static_cast<int32>(RandInt(0, input_vector.Dim() - output_vector->Dim()));
+  //2. get the successive expected_chunk_length * sample_frequency data.
+  output_vector->CopyFromVec(input_vector.Range(start_point, output_vector->Dim()));
+}
+
+void FvectorPerturb::AddNoise(MatrixBase<BaseFloat>* chunk) {
```

Comment:
These type of functions are very specific for your design. 
It is better to have generic function with noise and chunk as input to function!